File: frontend/src/lib/validations/health-goal.ts
import { z } from "zod";

/** Goal types — key + display label/emoji, one source for schema and UI. */
export const goalTypes = [
  { key: "lose_weight", label: "Lose weight", emoji: "📉" },
  { key: "gain_muscle", label: "Gain muscle", emoji: "💪" },
  { key: "maintain", label: "Maintain", emoji: "⚖️" },
  { key: "gain_weight", label: "Gain weight", emoji: "📈" },
  { key: "general_fitness", label: "General fitness", emoji: "🏃" },
] as const;

export type GoalType = (typeof goalTypes)[number]["key"];
export const goalTypeKeys = goalTypes.map((g) => g.key) as [GoalType, ...GoalType[]];

export const activityLevels = [
  { key: "sedentary", label: "Sedentary" },
  { key: "light", label: "Lightly active" },
  { key: "moderate", label: "Moderately active" },
  { key: "active", label: "Active" },
  { key: "very_active", label: "Very active" },
] as const;

export type ActivityLevel = (typeof activityLevels)[number]["key"];
export const activityLevelKeys = activityLevels.map((a) => a.key) as [
  ActivityLevel,
  ...ActivityLevel[],
];

/** The API contract for a health goal — mirrors the backend Pydantic schema. */
export const healthGoalSchema = z.object({
  goal_type: z.enum(goalTypeKeys),
  current_weight_kg: z.number().min(20).max(400).nullable(),
  target_weight_kg: z.number().min(20).max(400).nullable(),
  height_cm: z.number().min(50).max(260).nullable(),
  activity_level: z.enum(activityLevelKeys).nullable(),
  timeframe_weeks: z.number().int().min(1).max(520).nullable(),
  note: z.string().max(300).nullable(),
});

export type HealthGoalInput = z.infer<typeof healthGoalSchema>;

File: backend/app/schemas/health_goal.py
"""Request/response schemas for the user's health goal.

Mirrors the frontend Zod schema in `frontend/src/lib/validations/health-goal.ts`.
Only ``goal_type`` is required; the profile fields are optional and simply make
the AI daily-summary estimates more grounded.
"""

from datetime import datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field, field_validator

GoalType = Literal[
    "lose_weight", "gain_muscle", "maintain", "gain_weight", "general_fitness"
]
ActivityLevel = Literal["sedentary", "light", "moderate", "active", "very_active"]


class HealthGoalBase(BaseModel):
    goal_type: GoalType
    current_weight_kg: float | None = Field(default=None, ge=20, le=400)
    target_weight_kg: float | None = Field(default=None, ge=20, le=400)
    height_cm: float | None = Field(default=None, ge=50, le=260)
    activity_level: ActivityLevel | None = None
    timeframe_weeks: int | None = Field(default=None, ge=1, le=520)
    note: str | None = Field(default=None, max_length=300)

    @field_validator("note")
    @classmethod
    def _strip_note(cls, value: str | None) -> str | None:
        if value is None:
            return None
        stripped = value.strip()
        return stripped or None


class HealthGoalUpsert(HealthGoalBase):
    pass


class HealthGoalRead(HealthGoalBase):
    model_config = ConfigDict(from_attributes=True)

    id: int
    created_at: datetime
    updated_at: datetime
